### ISSUE TITLE
Anomaly: not a sort in match elaboration without return clause

### ISSUE THREAD

**Issue by eponier** [coq/coq#10917]
#### Description of the problem

The anomaly is produced on the following code:
```
Definition m (h : 0 = 1) P : P 0 -> P 1 :=
  fun H => match h, H with | eq_refl, _ => H end.
```
It disappears when a `return` clause is added, even the trivial `return _`.

```
Definition m (h : 0 = 1) P : P 0 -> P 1 :=
  fun H => match h, H return _ with | eq_refl, _ => H end.
```
It reminds me of another limitation of pattern-matching consisting in getting rejected because the wrong universe (namely `Prop`) was chosen in the absence of a `return` clause. Similarly to this issue, with a clause `return _` or `return Type`, the pattern-matching was accepted. I cannot find this issue at the moment.

#### Coq Version

8.7 and 8.10

**Comment by SkySkimmer** [comment]
Cases is generating the nonsensical predicate `fun n : nat => 0 = (S n) -> S n 0 -> Type` and sends it to typing.check_allowed_sort.
Maybe @herbelin knows what to do about it?

**Comment by herbelin** [comment]
@skyskimmer: thanks for the helpful analysis. See a fix at #11639.
